src/ccl_halomod.c

```diff
@@ -0,0 +1,355 @@
+  case constant:
+    
+    return 4.;
+
+    // Something went wrong
+  default:
+    
+    *status = CCL_ERROR_HALOCONC;
+    ccl_raise_exception(*status, "ccl_halomod.c: concentration-mass relation specified incorrectly");
+    return NAN;
+    	  
+  }
+}
+
+// Fourier Transforms of halo profiles
+static double window_function(ccl_cosmology *cosmo, double m, double k, double a, double odelta, ccl_win_label label, int *status){
+
+  double rho_matter, c;
+  
+  switch(label){
+
+  case NFW:
+
+    // The mean background matter density in Msun/Mpc^3
+    rho_matter = ccl_rho_x(cosmo, 1., ccl_species_m_label, 1, status);
+
+    // The halo concentration for this mass and scale factor  
+    c = ccl_halo_concentration(cosmo, m, a, odelta, Duffy2008_virial, status);    
+
+    // The function U is normalised so multiplying by M/rho turns units to overdensity
+    return m*u_nfw_c(cosmo, c, m, k, a, status)/rho_matter;
+
+    // Something went wrong
+  default:
+    
+    *status = CCL_ERROR_HALOWIN;
+    ccl_raise_exception(*status, "ccl_halomod.c: Window function specified incorrectly");
+    return NAN;
+    
+  }
+  
+}
+
+// Parameters structure for the one-halo integrand
+typedef struct{  
+  ccl_cosmology *cosmo;
+  double k, a;
+  int *status;
+} Int_one_halo_Par;
+
+// Integrand for the one-halo integral
+static double one_halo_integrand(double log10mass, void *params){  
+  
+  Int_one_halo_Par *p = (Int_one_halo_Par *)params;;
+  double halomass = pow(10,log10mass);
+  double Delta_v = Dv_BryanNorman(p->cosmo, p->a, p->status); // Virial density of haloes
+
+  // The squared normalised Fourier Transform of a halo profile (W(k->0 = 1)
+  double wk = window_function(p->cosmo,halomass, p->k, p->a, Delta_v, NFW, p->status);
+
+  // Fairly sure that there should be no ln(10) factor should be here since the integration is being specified in log10 range
+  double dn_dlogM = ccl_massfunc(p->cosmo, halomass, p->a, Delta_v, p->status);
+    
+  return dn_dlogM*pow(wk,2);
+}
+
+// The one-halo term integral using gsl
+static double one_halo_integral(ccl_cosmology *cosmo, double k, double a, int *status){
+
+  int one_halo_integral_status = 0, qagstatus;
+  double result = 0, eresult;
+  double log10mmin = log10(HM_MMIN);
+  double log10mmax = log10(HM_MMAX);
+  Int_one_halo_Par ipar;
+  gsl_function F;
+  gsl_integration_workspace *w = gsl_integration_workspace_alloc(1000);
+
+  // Structure required for the gsl integration
+  ipar.cosmo = cosmo;
+  ipar.k = k;
+  ipar.a = a;
+  ipar.status = &one_halo_integral_status;
+  F.function = &one_halo_integrand;
+  F.params = &ipar;
+
+  // Actually does the integration
+  qagstatus = gsl_integration_qag(&F, log10mmin, log10mmax, HM_EPSABS, HM_EPSREL, HM_LIMIT, HM_INT_METHOD, w, &result, &eresult);
+
+  // Clean up
+  gsl_integration_workspace_free(w);
+
+  // Check for errors
+  if (qagstatus != GSL_SUCCESS) {
+    ccl_raise_gsl_warning(qagstatus, "ccl_halomod.c: one_halo_integral:");
```

Comment:
Is this warning string complete? It ends with a colon.

Reply:
Okay, I changed this and made it similar to other gsl warnings that are raised in e.g., `ccl_power.c`. I followed the convention: